z3/src/ast.rs

```diff
@@ -236,6 +236,47 @@ pub trait Ast<'ctx>: Sized + fmt::Debug {
             )
         })
     }
+
+    /// Return the number of children of this AST node.
+    /// Leaf nodes (eg Bool consts) will return 0.
+    fn num_children(&self) -> usize {
+        let this_ctx = self.get_ctx().z3_ctx;
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let this_app = Z3_to_app(this_ctx, self.get_z3_ast());
+            Z3_get_app_num_args(this_ctx, this_app) as usize
+        }
+    }
+
+    /// Return the nth child of this AST node.  Out-of-bound indices will
+    /// return none.
+    fn nth_child(&self, idx: usize) -> Option<Dynamic<'ctx>> {
+        if idx >= self.num_children() {
+            None
+        } else {
+            let child_ast = unsafe { nth_child_ast_unchecked(self, idx) };
+            Some(Dynamic::new(self.get_ctx(), child_ast))
+        }
+    }
+
+    /// Return the children of this node as a Vec of Dynamic nodes.
+    fn children(&self) -> Vec<Dynamic<'ctx>> {
+        let n = self.num_children();
+        (0..n)
+            .map(|i| Dynamic::new(self.get_ctx(), unsafe { nth_child_ast_unchecked(self, i) }))
+            .collect()
+    }
+}
+
+// Retrieve the child at index `idx`.
+// This **does not check bounds**, the caller **must** ensure that
+// `idx < node.num_children()`.
+unsafe fn nth_child_ast_unchecked<'ctx>(node: &impl Ast<'ctx>, idx: usize) -> Z3_ast {
```

Comment:
I would prefer the simpler version where this function's body is inlined into `nth_child` and `children` calls `self.nth_child(i).unwrap()` unless there is a *really* good reason not to do so.

I would totally expect LLVM to easily determine that `i < self.num_children()` inside `children` and remove the unwrap checks from `self.nth_child(i).unwrap()`. Are you seeing otherwise? Also, the cost of the (very predictable) unwrap check should be dwarfed by the function call (especially if Z3 is in a shared library).